### PR TITLE
Empty string is a valid JSON-key

### LINKED ISSUE #4253: An empty string is a valid JSON attribute name
### Checks

* [ x] I added a descriptive title to this issue
* [ x] I have searched (google, github) for similar issues and couldn't find anything
* [ x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: False
                 install path: /Users/s_tsaplin/projects/github.com/sergeytsaplin/pydantic/pydantic
               python version: 3.9.13 (main, May 19 2022, 14:10:54)  [Clang 13.1.6 (clang-1316.0.21.2)]
                     platform: macOS-12.4-x86_64-i386-64bit
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```

```py
from pydantic import BaseModel, Field

class Model(BaseModel):
        empty_string_key: int = Field(alias='')

data = {'': 123}
m = Model(**data)
m.dict(by_alias=True)
```

## Current behavior:

The code causes the exception:

```
Traceback (most recent call last):
  File "<masked>/.pyenv/versions/3.10.1/lib/python3.10/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Model
empty_string_key
  field required (type=value_error.missing)
 ```

## Expected result:

The code should print:

```
{'': 123}
```

Links:
* https://stackoverflow.com/questions/58916957/is-an-empty-string-a-valid-json-key
* https://datatracker.ietf.org/doc/html/rfc6901#section-5

### PR DESCRIPTION
**Pull request by SergeyTsaplin**
## Change Summary

An empty string is a valid JSON attribute name, so aliases checking cannot be just like `bool(alias)` it should be strictly compared with `None`.

## Related issue number

fix #4253

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Thanks for the contribution.

While I think you're technically correct. I'm not sure I would support changing this at this late stage for pydantic V1.

I believe this is fixed in pydantic-core (the basis for pydantic V2) - if you're willing, it would be very useful to add unit tests for this to [pydantic-core](https://github.com/samuelcolvin/pydantic-core)?

[The plan](https://pydantic-docs.helpmanual.io/blog/pydantic-v2/) is to get v1.10 released with mostly cosmetic changes and some new features that having been hanging around for a long time, then move ahead with V2.

What do others think? @PrettyWood @hramezani

**Comment by hramezani** [comment]
Thanks @SergeyTsaplin for the patch :+1: and Thanks @samuelcolvin for explaining and asking my opinion.

> I'm not sure I would support changing this at this late stage for pydantic V1.

Why would be the problem? seems we don't need a big change to support it? 

> it would be very useful to add unit tests for this to [pydantic-core](https://github.com/samuelcolvin/pydantic-core)

:+1:

**Comment by samuelcolvin** [comment]
ye, I agree it's a small change. Just there's a lot in `fields.py` that scares me, its pretty complicated. My worry is subtle side-effects not covered tests, but maybe I'm being a wimp. :hand: 

Let's wait and see what @PrettyWood thinks.

**Comment by SergeyTsaplin** [comment]
@samuelcolvin I don't think that the change will break some existed code, it can break only when a user uses `Field(alias='')`  for not "aliased" fields (I have no idea why to do it:)), but of course, the final decision is up to you. For me, it's not a problem to wait for 1.11 if you accidentally decide to merge it to the V1.

About the unit-test to pydantic-core - I can do it, but not soon. It may take me a few days to find a time for it.

File: changes/4253-sergeytsaplin.md
Comment (samuelcolvin) [inline changes/4253-sergeytsaplin.md:1]: I think we can make this slightly clearer, something like

```suggestion
Allow empty string aliases by using a `alias is not None` check, rather than `bool(alias)`
```

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM

Let's go ahead with this.

**Comment by samuelcolvin** [comment]
> About the unit-test to pydantic-core - I can do it, but not soon. It may take me a few days to find a time for it.

No problem with a wait, this is a marathon, not a sprint.

**Comment by samuelcolvin** [comment]
no, just keep committing, we squash on merge.

**Review by PrettyWood — APPROVED** [review]
Looks good! I don't see any risk for this in v1.10

**Review by hramezani — APPROVED** [review]

**Comment by hramezani** [comment]
Let's wait for @samuelcolvin for final review and probably merge.

**Comment by samuelcolvin** [comment]
thanks so much.

**Comment by samuelcolvin** [comment]
Well (while I still think this is okay to be included in v1.10), my paranoia was well placed, tis was breaking starlite, see  https://github.com/starlite-api/starlite/pull/402.

Basically with pydantic v1's massive public (or semi-public) API, basically every change can break something. Bring on V2.

**Pull request merged**
